File: common/src/main/java/dev/obscuria/fragmentum/content/registry/Deferred.java
package dev.obscuria.fragmentum.content.registry;

import net.minecraft.core.Holder;

import java.util.function.Supplier;

public class Deferred<T> implements Supplier<T> {

    private final Holder<T> holder;

    public Deferred(Holder<T> holder) {
        this.holder = holder;
    }

    public Holder<T> holder() {
        return holder;
    }

    @Override
    public T get() {
        return holder.value();
    }
}

File: common/src/main/kotlin/dev/obscuria/fragmentum/registry/Deferred.kt
@file:Suppress("unused", "RedundantNullableReturnType", "UNCHECKED_CAST")

package dev.obscuria.fragmentum.registry

import net.minecraft.core.Holder

open class Deferred<T, V : T>(
    supplier: () -> Holder<T>
) : HolderProvider<T>, ValueProvider<V> {

    override val holder: Holder<T> by lazy { supplier.invoke() }
    override val value: V by lazy { holder.value() as V }
}